### DISCUSSION TITLE
Is it possible to serialize MX expressions

### DISCUSSION THREAD

**Discussion by lvanroye** [casadi/casadi#3516] [General]
Hello, 

I wonder whether it's possible to serialize MX expressions. Additionally, I wonder if I have two expressions e1 and e2 with a common subexpression, will this subexpression be duplicated? Like what would happen if you would make two seperate casadi::Functions, serialize them, deserialize them evaluate symbolically to retrieve the expressions. 

It would be very interesting if I could do something like the following (the commented code does not work but I included it to add the idea of what I'd like to do).

```
import casadi as cs

x = cs.MX.sym('x')

y = x
for i in range(10):
    y = cs.sin(y)

e1 = y
e2 = y + x


# x_s = cs.serialize(x)
# e1_s = cs.serialize(e1)
# e2_s = cs.serialize(e2)

# x_d = cs.deserialize(x_s)
# e1_d = cs.deserialize(e1_s)
# e2_d = cs.deserialize(e2_s)
```

**Comment by jgillis** [comment]
Yes, work with `ss = StringSerializer()`
```
ss.pack(x)
ss.pack(e1)
ss.pack(e2)

data = ss.encode()

ds = StringDeserializer(data)
x = ds.unpack()
e_1 = ds.unpack()
e_2 = ds.unpack()
```

**Reply by lvanroye** [reply]
That's great, thanks for the quick response!
